test/rule-matches/color-contrast-matches.js

```diff
@@ -22,216 +24,326 @@ describe('color-contrast-matches', function () {
+			var tree = axe._tree = axe.utils.getFlattenedTree(fixture);
+			assert.isTrue(rule.matches(shadowTarget, axe.utils.getNodeFromTree(tree[0], shadowTarget)));
+		});
+
+		it('should look at the correct root node when looking up an explicit label and disabled input', function () {
+			fixture.innerHTML = '<div id="parent">'+
+				'<input id="input">' +
+			'</div>';
+
+			var shadowRoot = document.getElementById('parent').attachShadow({ mode: 'open' });
+			shadowRoot.innerHTML = '<div id="shadowParent">' +
+					'<label for="input" id="shadowLabel">Label</label>' +
+					'<input id="input" disabled>' +
+				'</div>';
+
+			var shadowLabel = fixture.firstChild.shadowRoot.querySelector('#shadowLabel');
+			var tree = axe._tree = axe.utils.getFlattenedTree(fixture);
+			assert.isFalse(rule.matches(shadowLabel, axe.utils.getNodeFromTree(tree[0], shadowLabel)));
+		});
+
+		it('should look at the correct root node when looking up implicit label and disabled input', function () {
+			fixture.innerHTML = '<div id="parent">'+
+				'<input id="input">' +
+			'</div>';
+
+			var shadowRoot = document.getElementById('parent').attachShadow({ mode: 'open' });
+			shadowRoot.innerHTML = '<div id="shadowParent">' +
+					'<label for="input" id="shadowLabel">Label' +
+					'<input id="input" disabled></label>' +
+				'</div>';
+
+			var shadowLabel = fixture.firstChild.shadowRoot.querySelector('#shadowLabel');
+			var tree = axe._tree = axe.utils.getFlattenedTree(fixture);
+			assert.isFalse(rule.matches(shadowLabel, axe.utils.getNodeFromTree(tree[0], shadowLabel)));
+		});
+
+		it('should look at the correct root node for a disabled control\'s label associated w/ aria-labelledby', function () {
+			fixture.innerHTML = '<div id="parent">'+
+				'<input id="input">' +
+			'</div>';
+
+			var shadowRoot = document.getElementById('parent').attachShadow({ mode: 'open' });
+			shadowRoot.innerHTML = '<div id="shadowParent">' +
+					'<label id="shadowLabel">Label</label>' +
+					'<input aria-labelledby="shadowLabel" disabled>' +
+				'</div>';
+
+			var shadowLabel = fixture.firstChild.shadowRoot.querySelector('#shadowLabel');
+			var tree = axe._tree = axe.utils.getFlattenedTree(fixture);
+			assert.isFalse(rule.matches(shadowLabel, axe.utils.getNodeFromTree(tree[0], shadowLabel)));
+		});
+
+		it('should look at the children of a virtual node for overlap', function () {
+			fixture.innerHTML = '<div id="parent">'+
+				 '<div id="firstChild" style="background-color: #ccc; color: #fff;"></div>' +
+			'</div>';
+
+			var shadowRoot = document.getElementById('firstChild').attachShadow({ mode: 'open' });
+			shadowRoot.innerHTML = 'Some text' +
+					'<p style="color: #fff;" id="shadowTarget">Other text</p>';
+
+			var firstChild = fixture.querySelector('#firstChild');
+			var tree = axe._tree = axe.utils.getFlattenedTree(fixture);
+			assert.isTrue(rule.matches(firstChild, axe.utils.getNodeFromTree(tree[0], firstChild)));
+		});
+	}
```

Comment:
The approach I took with other tests was to put a second, enabled input in the light DOM with the same ID and an explicit pairing with the label–that way it would return `true` if the shadow boundary wasn't being handled properly. Otherwise it's difficult to test the `isFalse` case by itself since that would mean it didn't get matched (and could have occurred for any number of reasons)